app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java

```diff
@@ -0,0 +1,306 @@
+
+            // Get the device MAC address, which is the last 17 chars in the View
+            final String info = ((TextView) v).getText().toString();
+            final String address = info.substring(info.length() - 17);
+            final BluetoothDevice device = getDevice(address);
+            bluetoothConnectionService.connect(device);
+            try {
+                // busy wait is ugly, I think this needs the EventBus (see unused Handler in this class)
+                while (bluetoothConnectionService.getState() != BluetoothConnectionService.STATE_CONNECTED) { /* temporary */}
+                bluetoothConnectionService.writeAllContacts();
+            } catch (final IOException e) {
+                LOGGER.warn("Writing all contacts issued an IOexception");
+            }
+
+
+            // Create the result Intent and include the MAC address
+            final Intent intent = new Intent();
+            intent.putExtra(EXTRA_DEVICE_ADDRESS, address);
+
+
+            // Set result and finish this Activity
+            setResult(Activity.RESULT_OK, intent);
+            finish();
+
+        }
+    };
+    private ArrayAdapter<String> newDevicesArrayAdapter;
+    // Create a BroadcastReceiver for ACTION_FOUND.
+    private final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
+        public void onReceive(final Context context, final Intent intent) {
+            final String action = intent.getAction();
+            if (BluetoothDevice.ACTION_FOUND.equals(action)) {
+                final BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
+                // If it's already paired, skip it, because it's been listed already
+                if (device.getBondState() != BluetoothDevice.BOND_BONDED) {
+                    newDevices.add(device);
+                    newDevicesArrayAdapter.add(device.getName() + "\n" + device.getAddress());
+                }
+                // When discovery is finished, change the Activity title
+            } else if (BluetoothAdapter.ACTION_DISCOVERY_FINISHED.equals(action)) {
+                setTitle("select_device");
+                if (newDevicesArrayAdapter.getCount() == 0) {
+                    newDevicesArrayAdapter.add("none_found");
+                }
+            }
+
+        }
+    };
+    private ArrayAdapter<String> pairedDevicesArrayAdapter;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        // Register for broadcasts when a device is discovered.
+        IntentFilter filter = new IntentFilter(BluetoothDevice.ACTION_FOUND);
+        registerReceiver(broadcastReceiver, filter);
+
+        setContentView(R.layout.activity_bluetoothconection);
+
+        // Set result CANCELED in case the user backs out
+        setResult(Activity.RESULT_CANCELED);
+
+        // SetUp bluetooth adapter
+        setUp();
```

Comment:
Isn't the `onCreate` method intended for setup? Could you move the code from `setUp()` to this method?